### DISCUSSION TITLE
Custom metadata over service call

### DISCUSSION THREAD

**Discussion by avdmr** [music-assistant/support#2006] [Feature requests and ideas]
First off, let me just say how much I like Music Assistant. It's seriously awesome. Big kudos to the team for such an amazing integration.
I'm currently trying to create a custom service integration that takes a YouTube video URL, extracts the audio stream link and plays it on a media player.  I've been trying to send custom metadata via the advanced service call, following the instructions from Home Assistant. But, it seems like the standard method just isn't doing the trick.

Let me give you an example of what I'm talking about:
```service: mass.play_media
data:
  media_id: "url to music"
  media_type: track
  extra:
    metadataType: 3
    title: custom title
    artist: test
    thumb: "https://brands.home-assistant.io/_/homeassistant/logo.png"
target:
  entity_id: media_player.speaker_1
```

at the moment it seems to use a part of the URL as the title.

![afbeelding](https://github.com/music-assistant/support/assets/89525698/46597448-5cf5-4c92-9c78-d659ea90cd14)

So, I was wondering if there's any chance we could make it possible to send custom metadata through the advanced service call feature? I think it would be nice feature to have. I totally get that this might take some time and effort, but I reckon it'd be a interesting feature.

**Comment by SteveDinn** [comment]
Why not just give back all the metadata that the provider can offer? When I search for an album in Home Assistant (using the `music_assistant.get_library` action
```
action: music_assistant.get_library
data:
  config_entry_id: 01JRTR5CRSQ6MK7RK35FK63XV0
  media_type: album
  album_artists_only: false
  limit: 25
  offset: 0
  order_by: timestamp_added_desc
  search: nevermind
```

I only get the following information. It isn't much:
```
  - media_type: album
    uri: library://album/2397
    name: Nevermind
    version: ""
    image: >-
      http://10.1.1.9:8095/imageproxy?path=al-6Nt8bloRR9mxSYNbWf3O0v_6405edfd&provider=opensubsonic--3RRKArfV&size=0
    artists:
      - media_type: artist
        uri: library://artist/739
        name: Nirvana
        version: ""
        image: null
```

With the increasing availablility of LLMs and voice commands and whatever else can invoke the playing of music, it would be nice to be able to specify the search by whatever metadata is available. I know there is [another feature request for genre](https://github.com/orgs/music-assistant/discussions/562), but what about things like:
- **Date added** (to the source provider library, not music assistant's library)
- **Release date**
- **Original release date** (if the album is a re-issue)
- **Track Length** (or **Album Length**)
- **Quality** (encoding method, sample size & frequency)
- **Number of channels**

I'm sure there are many more bits of metadata available from most providers that I just can't think of off the top of my head. I could then specify to my LLM (or explicitly in a script): _"Shuffle the 5 most recently-added albums"_ or _"Play albums by Nirvana in order of release date"_ or _"Play New Found Glory's third album"_. Above and beyond the _genre_ issue, the more metadata there is available for automations, the better automations we can create.

For individual tracks, here are some of the bits of metadata that [beets](https://beets.io) applies to my music files when I add them to my library. Obviously, I don't think _everything_ is necessary, but it's nice to have, and there are certain bits of data that would make fantastic "quality of life" improvements for users.

Input #0, flac, from '08 - Broken Heart.flac':
| Property                  | Value                                                                    |
|---------------------------|--------------------------------------------------------------------------|
| acoustid_fingerprint      | AQADtE4kiVKiBX-CJ_nR_Ei2ZIlyRMrQ73iSxcFtvNGR8Ed0PEeZiDnO7...             |
| acoustid_id               | 1fbb50bb-1ada-4215-8973-56f405d2bbdc                                     |
| album                     | Even If It Kills Me                                                      |
| album artist              | Motion City Soundtrack                                                   |
| album_artist              | Motion City Soundtrack                                                   |
| album_artists             | Motion City Soundtrack                                                   |
| albumartists              | Motion City Soundtrack                                                   |
| albumartists_credit       | Motion City Soundtrack                                                   |
| albumartists_sort         | Motion City Soundtrack                                                   |
| albumartistsort           | Motion City Soundtrack                                                   |
| artist                    | Motion City Soundtrack                                                   |
| artist_credit             | Motion City Soundtrack                                                   |
| artists                   | Motion City Soundtrack                                                   |
| artists_credit            | Motion City Soundtrack                                                   |
| artists_sort              | Motion City Soundtrack                                                   |
| artistsort                | Motion City Soundtrack                                                   |
| barcode                   | 045778686261                                                             |
| bpm                       | 0                                                                        |
| compilation               | 0                                                                        |
| composer                  | Joshua Cain, Jesse Johnson, Justin Courtney Pierre, Matthew Taylor, Tony Thaxton |
| composersort              | Cain, Joshua, Johnson, Jesse, Pierre, Justin Courtney, Taylor, Matthew, Thaxton, Tony |
| date                      | 2007-09-18                                                               |
| disc                      | 1                                                                        |
| discc                     | 1                                                                        |
| disctotal                 | 1                                                                        |
| encodedby                 | Lavf59.27.100                                                            |
| encoder                   | Lavf59.27.100                                                            |
| genre                     | Pop Punk                                                                 |
| isrc                      | USEP40715101                                                             |
| label                     | Epitaph                                                                  |
| lyricist                  | Joshua Cain, Jesse Johnson, Justin Courtney Pierre, Matthew Taylor, Tony Thaxton |
| media                     | Digital Media                                                            |
| musicbrainz_albumartistid | 190e78ac-8f44-4bf1-a4ff-95ed9376ad9e                                     |
| musicbrainz_albumid       | ad62a326-e9b5-4847-8349-19257b2a0b8d                                     |
| musicbrainz_albumstatus   | Official                                                                 |
| musicbrainz_albumtype     | album                                                                    |
| musicbrainz_artistid      | 190e78ac-8f44-4bf1-a4ff-95ed9376ad9e                                     |
| musicbrainz_releasegroupid| cce7c4c8-6577-387a-b062-3fff4490f399                                     |
| musicbrainz_releasetrackid| 82ce766f-3ab7-4c91-85ec-d795410245f2                                     |
| musicbrainz_trackid       | 108a1d92-7883-4697-97b0-8e54a4f9aaef                                     |
| musicbrainz_workid        | ea6e4d04-d63e-4c71-a0d1-0b3ccf6aef48                                     |
| originaldate              | 2007-07-17                                                               |
| publisher                 | Epitaph                                                                  |
| releasecountry            | XW                                                                       |
| releasestatus             | Official                                                                 |
| releasetype               | album                                                                    |
| script                    | Latn                                                                     |
| title                     | Broken Heart                                                             |
| totaldiscs                | 1                                                                        |
| totaltracks               | 13                                                                       |
| track                     | 8                                                                        |
| trackc                    | 13                                                                       |
| tracktotal                | 13                                                                       |
| year                      | 2014                                                                     |
